lib/widgets/katex.dart

```diff
@@ -0,0 +1,202 @@
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/widgets.dart';
+import 'package:flutter/rendering.dart';
+
+class NegativeLeftOffset extends SingleChildRenderObjectWidget {
+  NegativeLeftOffset({super.key, required this.leftOffset, super.child})
+    : assert(leftOffset.isNegative),
+      _padding = EdgeInsets.only(left: leftOffset);
+
+  final double leftOffset;
+  final EdgeInsetsGeometry _padding;
+
+  @override
+  RenderNegativePadding createRenderObject(BuildContext context) {
+    return RenderNegativePadding(
+      padding: _padding,
+      textDirection: Directionality.maybeOf(context));
+  }
+
+  @override
+  void updateRenderObject(
+    BuildContext context,
+    RenderNegativePadding renderObject,
+  ) {
+    renderObject
+      ..padding = _padding
+      ..textDirection = Directionality.maybeOf(context);
+  }
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<EdgeInsetsGeometry>('padding', _padding));
+  }
+}
+
+// Like [RenderPadding] but supports negative values.
+class RenderNegativePadding extends RenderShiftedBox {
```

Comment:
Let's leave a TODO tracking our intention to get a version of this upstream:
```suggestion
// Like [RenderPadding] but supports negative values.
// TODO(upstream): give Padding an option to accept negative padding (at cost of hit-testing not working)
class RenderNegativePadding extends RenderShiftedBox {
```